### ISSUE TITLE
One CPU fully occupied by Datadog when device offline

### ISSUE THREAD

**Issue by torbjorn-humane** [DataDog/dd-sdk-android#1469]
We're using the Datadog android java SDK to collect both traces and logs on an android device. The device may be offline for an extended period, which means logs/traces cannot be uploaded but are stored on device. After a while in this state, the datadog agent is fully occupying 1 CPU. We have reproduced the issue both on device and using the android emulator. We have tested both version 1.9.0 and 1.19.2 and the issue persists. We're using the Datadog-Timber integration for log/trace capturing.

**Steps to reproduce the issue:**

1. Create an android instrumented test that loops over a log statement say 100k times. 
2. Start the android emulator and enable airplane mode on the emulated device.
3. Run the test and observe the folder on device (using e.g. adb shell) where the cached files are stored (e.g. `/data/user/0/<myapplication>/cache/datadog-.../logs-v2`
4. When the number of cached files reaches ~200, stop the test.
5. Create another instrumented test that logs in a loop. For each iteration of the loop, sleep 10ms (emulating 100 log entries/second).
6. Run the second test and attach the profiler, e.g. android studio. One of the worker threads will occupy one CPU 100%. Breaking in the debugger indicates that the Datadog agent is using one of the worker threads to iterate over every stored file repeatedly, causing the high CPU usage.

**Describe what you expected:**
Running the same test with datadog enabled, emulating 100 log writes/second, occupies ~8% of one CPU when the device has network access. I'd expect no more CPU usage even if the device is offline.

**Additional context**
 - Android OS version: 31
 - Emulator: Pixel 6 Pro
 - Datadog SDK version: 1.9.0, 1.19.2
 - Versions of any other relevant dependencies (OkHttp, …): Timber 5.0.1, okhttp 4.9.3
 - Proguard configuration: N/A
 - Gradle Plugins: N/A

**Comment by torbjorn-humane** [comment]
Hi and thanks for getting back to me! You're of course right, my example is doing some excessive logging but it's mostly a quick way to demonstrate the problem. Our real logging rate is also somewhat excessive but not like above. Still, because the device can be offline for hours and hours the problem occurs eventually. Below are a couple of callstacks I get from the worker thread when breaking the test:

Example 1
```
isFileSafe:85, FileExtKt (com.datadog.android.core.internal.persistence.file)
accept:285, BatchFileOrchestrator$BatchFileFilter (com.datadog.android.core.internal.persistence.file.batch)
listFiles:1298, File (java.io)
invoke:108, FileExtKt$listFilesSafe$2 (com.datadog.android.core.internal.persistence.file)
invoke:106, FileExtKt$listFilesSafe$2 (com.datadog.android.core.internal.persistence.file)
safeCall:36, FileExtKt (com.datadog.android.core.internal.persistence.file)
listFilesSafe:106, FileExtKt (com.datadog.android.core.internal.persistence.file)
listSortedBatchFiles:272, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
freeSpaceIfNeeded:238, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
getWritableFile:51, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
writeCurrentBatch$lambda-0:57, ConsentAwareStorage (com.datadog.android.v2.core.internal.storage)
$r8$lambda$9ONtzLqLkkJDgpJ2t6E_ex_POv4:-1, ConsentAwareStorage (com.datadog.android.v2.core.internal.storage)
run:-1, ConsentAwareStorage$$ExternalSyntheticLambda0 (com.datadog.android.v2.core.internal.storage)
call:462, Executors$RunnableAdapter (java.util.concurrent)
run:266, FutureTask (java.util.concurrent)
runWorker:1167, ThreadPoolExecutor (java.util.concurrent)
run:641, ThreadPoolExecutor$Worker (java.util.concurrent)
run:920, Thread (java.lang)
```

Example 2
```
digit:6826, Character (java.lang)
digitOf:306, CharsKt__CharJVMKt (kotlin.text)
toLongOrNull:177, StringsKt__StringNumberConversionsKt (kotlin.text)
toLongOrNull:132, StringsKt__StringNumberConversionsKt (kotlin.text)
invoke:228, BatchFileOrchestrator$deleteObsoleteFiles$1 (com.datadog.android.core.internal.persistence.file.batch)
invoke:228, BatchFileOrchestrator$deleteObsoleteFiles$1 (com.datadog.android.core.internal.persistence.file.batch)
calcNext:171, FilteringSequence$iterator$1 (kotlin.sequences)
hasNext:194, FilteringSequence$iterator$1 (kotlin.sequences)
forEach:1290, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
deleteObsoleteFiles:229, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
getWritableFile:50, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
writeCurrentBatch$lambda-0:57, ConsentAwareStorage (com.datadog.android.v2.core.internal.storage)
$r8$lambda$9ONtzLqLkkJDgpJ2t6E_ex_POv4:-1, ConsentAwareStorage (com.datadog.android.v2.core.internal.storage)
run:-1, ConsentAwareStorage$$ExternalSyntheticLambda0 (com.datadog.android.v2.core.internal.storage)
call:462, Executors$RunnableAdapter (java.util.concurrent)
run:266, FutureTask (java.util.concurrent)
runWorker:1167, ThreadPoolExecutor (java.util.concurrent)
run:641, ThreadPoolExecutor$Worker (java.util.concurrent)
run:920, Thread (java.lang)
```
Example 3
```
reset:1164, Matcher (java.util.regex)
reset:256, Matcher (java.util.regex)
<init>:167, Matcher (java.util.regex)
matcher:1030, Pattern (java.util.regex)
matches:107, Regex (kotlin.text)
accept:286, BatchFileOrchestrator$BatchFileFilter (com.datadog.android.core.internal.persistence.file.batch)
listFiles:1298, File (java.io)
invoke:108, FileExtKt$listFilesSafe$2 (com.datadog.android.core.internal.persistence.file)
invoke:106, FileExtKt$listFilesSafe$2 (com.datadog.android.core.internal.persistence.file)
safeCall:36, FileExtKt (com.datadog.android.core.internal.persistence.file)
listFilesSafe:106, FileExtKt (com.datadog.android.core.internal.persistence.file)
listSortedBatchFiles:272, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
getReusableWritableFile:191, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
getWritableFile:54, BatchFileOrchestrator (com.datadog.android.core.internal.persistence.file.batch)
writeCurrentBatch$lambda-0:57, ConsentAwareStorage (com.datadog.android.v2.core.internal.storage)
$r8$lambda$9ONtzLqLkkJDgpJ2t6E_ex_POv4:-1, ConsentAwareStorage (com.datadog.android.v2.core.internal.storage)
run:-1, ConsentAwareStorage$$ExternalSyntheticLambda0 (com.datadog.android.v2.core.internal.storage)
call:462, Executors$RunnableAdapter (java.util.concurrent)
run:266, FutureTask (java.util.concurrent)
runWorker:1167, ThreadPoolExecutor (java.util.concurrent)
run:641, ThreadPoolExecutor$Worker (java.util.concurrent)
run:920, Thread (java.lang)
```

**Comment by 0xnm** [comment]
Hello @torbjorn-humane. Thanks for details!

The issue comes from the fact that syscall done by `File.isFile` is quite expensive, and regex check on the 200+ files also adds up. The picture with CPU consumption will be the same even if there is network, the problem is that there are many batch files to go through in the internal event write logic (so that if write frequency dominates over upload capabilities and the overall file queue is big - it will be the same).

We can probably optimize this hot path by throwing in some cache to avoid repetitive syscalls/regex matching for the known entities, we will try to apply the fix soon.

**Comment by torbjorn-humane** [comment]
Hi! Any news on this one? Anything I can help with?

**Comment by 0xnm** [comment]
Hello @torbjorn-humane! Sorry, for the delay. Can you please help us to estimate the priority of this issue, is it a blocker for you to adopt our SDK?

**Comment by torbjorn-humane** [comment]
Hi, thanks for getting back to me! Yeah, for the moment we've disabled the datadog agent completely across all our devices so not to disrupt users, but we're taking a hit on the debuggability for sure. A fix would be very much appreciated!

**Comment by 0xnm** [comment]
Hello @torbjorn-humane! We've made a quick improvement in https://github.com/DataDog/dd-sdk-android/pull/1506, hope it will help in your scenario.

Ideally we should change the logic of working with files to address your scenario better, but it will be a bigger change which goes outside simple patch release. We have the necessary task in our backlog, but I cannot give you any estimates.

We will probably release a new patch version `1.19.3` early next week with the change I've mentioned.

Worth to add that since the issue is coming from the number of files stored in the filesystem, you can also try to change batch size option to use bigger batch size (`setBatchSize`) in the SDK configuration in order to reduce the average number of batch files.

Thanks for your feedback with such scenario, we are always happy to hear how SDK behaves in "non-optimistic" cases. Can you please also share a bit more details on the application usage? Normally when application is very active interacting with SDK and may have no network connection for a long time, it is probably some PoS/Kiosk which is on external power source, so even if CPU usage is higher than expected, it doesn't cause a big issue.

**Comment by torbjorn-humane** [comment]
Thank you @0xnm, this is very much appreciated! I will pick up the new release as soon as it's available and verify on our end before upgrading our software.

I unfortunately can't share anything about our use case that isn't [already public](https://www.google.com/search?q=humane+device&ei=aiemZI32IpbNkPIP__Ot0A8&ved=0ahUKEwiN08f3hPn_AhWWJkQIHf95C_oQ4dUDCBA&uact=5&oq=humane+device&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQAzIICAAQgAQQsQMyBQgAEIAEMgUIABCABDIFCAAQgAQyBQgAEIAEMgUIABCABDIFCAAQgAQyBQgAEIAEMggIABCKBRCGAzIICAAQigUQhgM6BAgAEEc6CwgAEIAEELEDEIMBOgsILhCABBDHARCvAToICAAQgAQQyQM6CAgAEIoFEJIDOgsIABCKBRCxAxCDAToRCC4QigUQsQMQgwEQxwEQ0QM6BwgAEIoFEEM6BwgAEIAEEApKBAhBGABQ6gNYnQtg-QtoAHACeACAAWGIAZYEkgEBNpgBAKABAcABAcgBCA&sclient=gws-wiz-serp), but stay tuned. :)

**Comment by 0xnm** [comment]
Hello @torbjorn-humane! We've released version [1.19.3](https://github.com/DataDog/dd-sdk-android/releases/tag/1.19.3) with the optimization for your use-case included. Please check it out and let us know if it unblocks you.

Meanwhile, if you are in the process of the initial Datadog SDK adoption, you may be interested in our [first beta of SDK v2](https://github.com/DataDog/dd-sdk-android/releases/tag/2.0.0-beta1) (note that artifact names are different now), which gives more granularity in comparison with SDK v1 (say if you need only Logs, you may now only import `logs` module without the overhead of RUM and Trace classes in memory). This beta release includes the optimization mentioned above as well.

**Comment by 0xnm** [comment]
Hello @torbjorn-humane! I'm closing this issue due to inactivity, but if you still experience the problem feel free to re-open it.
